man/ibccquery.8
.TH IBCCQUERY 8 "Aug 10, 2011" "OpenIB" "OpenIB Diagnostics"

.SH NAME
ibccquery \- query congestion control settings/info

.SH SYNOPSIS
.B ibccquery
[\-d(ebug)] [\-G(uid)] [\-C ca_name] [\-P ca_port] [\-t(imeout) timeout_ms]
[\-V(ersion)] [\-h(elp)] [\-c cckey] <op> <lid|guid> [port]

.SH DESCRIPTION
.PP
.Bibccquery support the querying of settings and other information related
to congestion control.

.SH OPTIONS

.PP
.TP
Current supported operations and their parameters:
  CongestionInfo (CI) <addr>
  CongestionKeyInfo (CK) <addr>
  CongestionLog (CL) <addr>
  SwitchCongestionSetting (SS) <addr>
  SwitchPortCongestionSetting (SP) <addr> [<portnum>]
  CACongestionSetting (CS) <addr>
  CongestionControlTable (CT) <addr>
  Timestamp (TI) <addr>

.TP
\fB\-c\-fR, \fB\-\-cckey\fR <cckey>
Specify a congestion control (CC) key.  If none is specified, a key of 0 is used.

.SH COMMON OPTIONS

Most OpenIB diagnostics take the following common flags. The exact list of
supported flags per utility can be found in the usage message and can be shown
using the util_name -h syntax.

# Debugging flags
.PP
\-d      raise the IB debugging level.
        May be used several times (-ddd or -d -d -d).
.PP
\-e      show send and receive errors (timeouts and others)
.PP
\-h      show the usage message
.PP
\-v      increase the application verbosity level.
        May be used several times (-vv or -v -v -v)
.PP
\-V      show the version info.

# Addressing flags
.PP
\-G      use GUID address argument. In most cases, it is the Port GUID.
        Example:
        "0x08f1040023"
.PP
\-s <smlid>      use 'smlid' as the target lid for SM/SA queries.

# Other common flags:
.PP
\-C <ca_name>    use the specified ca_name.
.PP
\-P <ca_port>    use the specified ca_port.
.PP
\-t <timeout_ms> override the default timeout for the solicited mads.

If a port and/or CA name is specified, the user request is
attempted to be fulfilled, and will fail if it is not possible.

.SH EXAMPLES

.PP
ibccquery CongestionInfo 3		# Congestion Info by lid
.PP
ibccquery SwitchPortCongestionSetting 3	# Query all Switch Port Congestion Settings
.PP
ibccquery SwitchPortCongestionSetting 3 1 # Query Switch Port Congestion Setting for port 1

.SH AUTHOR
.TP
Albert Chu
.RI < chu11@llnl.gov >


ChangeLog
commit 119aabce5e6722336e9f5a97cea14eebab2376ca
Author: Steve Wise <swise@opengridcomputing.com>
Date:   Wed Nov 9 09:50:22 2011 -0600

    Spin release 1.2.0.
    
    Signed-off-by: Steve Wise <swise@opengridcomputing.com>

commit 721cfb11ca5697d017bc60167805aea9bbb13ba0
Author: Vipul Pandya <vipul@chelsio.com>
Date:   Mon Nov 7 15:08:45 2011 +0530

    libcxgb4: Reset flush_cidx before incrementing cidx.
    
    Signed-off-by: Vipul Pandya <vipul@chelsio.com>

commit db4fe38c7ee2b3ad9dc8c33f2b54f0c8cdf8b4d2
Author: Vipul Pandya <vipul@chelsio.com>
Date:   Mon Nov 7 15:06:29 2011 +0530

    libcxgb4: Fix multi-flush logic path....again
    
    Signed-off-by: Vipul Pandya <vipul@chelsio.com>

commit 19198bdc2f7feb05babcc2a544ae8b48e4943c41
Author: Vipul Pandya <vipul@chelsio.com>
Date:   Mon Nov 7 15:02:57 2011 +0530

    libcxgb4: Reset cidx_flush if a active queue starts getting hw cqes
    
    Signed-off-by: Vipul Pandya <vipul@chelsio.com>

commit 98545f3a8cbe1670efad080cf908db42269f76ff
Author: Vipul Pandya <vipul@chelsio.com>
Date:   Mon Nov 7 14:56:44 2011 +0530

    libcxgb4: Support for flushing a CQ bound to multiple QPs
    
    Signed-off-by: Vipul Pandya <vipul@chelsio.com>

commit 86ba968706b3dc8cb232a303201bb03a2d919c88
Author: Vipul Pandya <vipul@chelsio.com>
Date:   Mon Nov 7 13:06:43 2011 +0530

    libcxgb4: Checks iw_cxgb4 ABI version to know door bell drop recovery support
    
    Store iw_cxgb4 ABI version in a global state. Read that global state in
    t4_wq_db_enabled to know if iw_cxgb4 driver supports door bell drop recovery.
    
    Signed-off-by: Vipul Pandya <vipul@chelsio.com>

commit 2b6563ed5b377dac56c218b809d387089d70312d
Author: Vipul Pandya <vipul@chelsio.com>
Date:   Sun Oct 23 16:50:16 2011 +0530

    libcxgb4: fixed compilation warnings
    
    Signed-off-by: Vipul Pandya <vipul@chelsio.com>

commit c5c120969c22a04d34adc566085192afc99d0db9
Author: Vipul Pandya <vipul@chelsio.com>
Date:   Fri Oct 21 19:05:20 2011 +0530

    libcxgb4: Bug Fix#4410 - Fix 3 segfaults
    
    This patch fixes 3 segfaults with below signatures:
    ----
    1)
    Program received signal SIGSEGV, Segmentation fault.
    [Switching to Thread 0x7ffff7fef710 (LWP 22624)]
    0x00007ffff75613e4 in create_read_req_cqe (wq=0x7fffe800ada0,
    hw_cqe=0x7ffff7fc7880, read_cqe=0x7ffff7fee860) at src/cq.c:162
    162             read_cqe->u.rdma.u.scqe.cidx = wq->sq.oldest_read->idx;
    (gdb) bt
    hw_cqe=0x7ffff7fc7880, read_cqe=0x7ffff7fee860) at src/cq.c:162
    attr=0x7ffff7feeae0, attr_mask=1) at src/verbs.c:691
    src/verbs.c:721
    
    2)
    Program received signal SIGSEGV, Segmentation fault.
    [Switching to Thread 0x7ffff6e07710 (LWP 22218)]
    0x00007ffff71563e4 in create_read_req_cqe (wq=0x7fffe800ada0,
    hw_cqe=0x7ffff7fc7880, read_cqe=0x7ffff6e02b70) at src/cq.c:162
    162             read_cqe->u.rdma.u.scqe.cidx = wq->sq.oldest_read->idx;
    (gdb) bt
    hw_cqe=0x7ffff7fc7880, read_cqe=0x7ffff6e02b70) at src/cq.c:162
    src/verbs.c:887
    
    3)
    (gdb) bt
    wc=0x7ffff51378d0) at src/cq.c:776
    evd_ptr=<value optimized out>, wc_ptr=<value optimized out>)
        at /usr/include/infiniband/verbs.h:957
    ----
    
    Signed-off-by: Vipul Pandya <vipul@chelsio.com>

commit 075adec1159606303b40f4b29ef3cea7ab780ac2
Author: Vipul Pandya <vipul@chelsio.com>
Date:   Fri Oct 21 18:59:16 2011 +0530

    libcxgb4: return -ENOSYS in c4iw_resize_cq().
    
    I fixed a "bug" in c4iw_resize_cq() in an earlier commit which regressed
    us.  c4iw_resize_cq() was returning 0 always even though we don't support
    resizing the CQ.  So I change libcxgb4 to return whatever errno is
    returned by the kernel driver.   However, because the driver didn't set
    the uverbs capabilities for resize_cq, uverbs returned EINVAL.  The OMPI
    code checks for -ENOSYS and handles providers that don't support it.
    If, however, ibv_resize_cq() returns any other error, OMPI bails out.
    So we need to return -ENOSYS.
    
    I didn't see this error, however, running NP192 OMPI/IMB all night long.
    So it must only try to resize for small jobs and thus I missed the
    regression.
    
    Signed-off-by: Vipul Pandya <vipul@chelsio.com>

commit 320285a3a3a93a614e24a78f7a68780d7791e3c5
Author: Vipul Pandya <vipul@chelsio.com>
Date:   Fri Oct 21 18:53:47 2011 +0530

    libcxgb4: Don't flush unsignaled WRs if the QP closed normally
    
    - add query_qp support in driver to query the qp state before flushing.
    
    - fixed flush logic (again).
    
    - don't flush unsignaled WRs is QP had a normal close.
    
    - support up to 64K QIDs.
    
    Signed-off-by: Vipul Pandya <vipul@chelsio.com>

commit 4c98bebe97aeef1c5cda2f071eb9c1d4fa29c37e
Author: Vipul Pandya <vipul@chelsio.com>
Date:   Fri Oct 21 17:58:38 2011 +0530

    libcxgb4: Correctly flush unsignaled WRs followed by a signaled WR.
    
    The cq flush logic was not accounting for unsignaled SQ WRs followed by
    a signaled WR.  The result was too many flushed CQEs causing application
    faults or assert() failures in libcxgb4 because the SQ state is corrupted.
    
    Signed-off-by: Vipul Pandya <vipul@chelsio.com>

commit 8196f85bdefb1511d6d834240a92fdaf815834b8
Author: Vipul Pandya <vipul@chelsio.com>
Date:   Fri Oct 21 17:31:03 2011 +0530

    libcxgb4: Remove auto generated ltmain.sh file.
    
    Signed-off-by: Vipul Pandya <vipul@chelsio.com>

commit 230d12f42d143eeb53d8c4935ba27620c13e6165
Author: Vipul Pandya <vipul@chelsio.com>
Date:   Fri Oct 21 17:20:37 2011 +0530

    libcxgb4: Add debug code to dump state on a stall condition.
    
    Signed-off-by: Vipul Pandya <vipul@chelsio.com>

commit 2df0677b22aebb7a22d0e8ca11beb90250b363c7
Author: Vipul Pandya <vipul@chelsio.com>
Date:   Fri Oct 21 15:04:46 2011 +0530

    libcxgb4: don't write host idx values when wq is in error.
    
    This fixes a regression introduced by the db recovery changes.  The host

src/ibccquery.c
/*
 * Copyright (c) 2004-2009 Voltaire Inc.  All rights reserved.
 * Copyright (c) 2011 Mellanox Technologies LTD.  All rights reserved.
 * Copyright (c) 2011 Lawrence Livermore National Lab.  All rights reserved.
 *
 * This software is available to you under a choice of one of two
 * licenses.  You may choose to be licensed under the terms of the GNU
 * General Public License (GPL) Version 2, available from the file
 * COPYING in the main directory of this source tree, or the
 * OpenIB.org BSD license below:
 *
 *     Redistribution and use in source and binary forms, with or
 *     without modification, are permitted provided that the following
 *     conditions are met:
 *
 *      - Redistributions of source code must retain the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer.
 *
 *      - Redistributions in binary form must reproduce the above
 *        copyright notice, this list of conditions and the following
 *        disclaimer in the documentation and/or other materials
 *        provided with the distribution.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
 * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

#if HAVE_CONFIG_H
#  include <config.h>
#endif				/* HAVE_CONFIG_H */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <getopt.h>
#include <netinet/in.h>

#define __STDC_FORMAT_MACROS
#include <inttypes.h>

#include <infiniband/mad.h>

#include "ibdiag_common.h"

struct ibmad_port *srcport;

typedef char *(op_fn_t) (ib_portid_t * dest, char **argv, int argc);

typedef struct match_rec {
	const char *name, *alias;
	op_fn_t *fn;
	unsigned opt_portnum;
} match_rec_t;

static op_fn_t class_port_info;
static op_fn_t congestion_info;
static op_fn_t congestion_key_info;
static op_fn_t congestion_log;
static op_fn_t switch_congestion_setting;
static op_fn_t switch_port_congestion_setting;
static op_fn_t ca_congestion_setting;
static op_fn_t congestion_control_table;
static op_fn_t timestamp_dump;

static const match_rec_t match_tbl[] = {
	{"ClassPortInfo", "CP", class_port_info, 0},
	{"CongestionInfo", "CI", congestion_info, 0},
	{"CongestionKeyInfo", "CK", congestion_key_info, 0},
	{"CongestionLog", "CL", congestion_log, 0},
	{"SwitchCongestionSetting", "SS", switch_congestion_setting, 0},
	{"SwitchPortCongestionSetting", "SP", switch_port_congestion_setting, 1},
	{"CACongestionSetting", "CS", ca_congestion_setting, 0},
	{"CongestionControlTable", "CT", congestion_control_table, 0},
	{"Timestamp", "TI", timestamp_dump, 0},
	{0}
};

uint64_t cckey = 0;

/*******************************************/
static char *class_port_info(ib_portid_t * dest, char **argv, int argc)
{
	char buf[2048];
	char data[IB_CC_DATA_SZ] = { 0 };

	if (!cc_query_status_via(data, dest, CLASS_PORT_INFO,
				 0, 0, NULL, srcport, cckey))
		return "class port info query failed";

	mad_dump_classportinfo(buf, sizeof buf, data, sizeof data);

	printf("# ClassPortInfo: %s\n%s", portid2str(dest), buf);
	return NULL;
}

static char *congestion_info(ib_portid_t * dest, char **argv, int argc)
{
	char buf[2048];
	char data[IB_CC_DATA_SZ] = { 0 };

	if (!cc_query_status_via(data, dest, IB_CC_ATTR_CONGESTION_INFO,
				 0, 0, NULL, srcport, cckey))
		return "congestion info query failed";

	mad_dump_cc_congestioninfo(buf, sizeof buf, data, sizeof data);

	printf("# CongestionInfo: %s\n%s", portid2str(dest), buf);
	return NULL;
}

static char *congestion_key_info(ib_portid_t * dest, char **argv, int argc)
{
	char buf[2048];
	char data[IB_CC_DATA_SZ] = { 0 };
	
	if (!cc_query_status_via(data, dest, IB_CC_ATTR_CONGESTION_KEY_INFO,
				 0, 0, NULL, srcport, cckey))
		return "congestion key info query failed";

	mad_dump_cc_congestionkeyinfo(buf, sizeof buf, data, sizeof data);

	printf("# CongestionKeyInfo: %s\n%s", portid2str(dest), buf);
	return NULL;
}

static char *congestion_log(ib_portid_t * dest, char **argv, int argc)
{
	char buf[2048];
	char data[IB_CC_LOG_DATA_SZ] = { 0 };
	char emptybuf[16] = { 0 };
	int i, type;

	if (!cc_query_status_via(data, dest, IB_CC_ATTR_CONGESTION_LOG,
				 0, 0, NULL, srcport, cckey))
		return "congestion log query failed";

	mad_decode_field((uint8_t *)data, IB_CC_CONGESTION_LOG_LOGTYPE_F, &type);

	if (type != 1 && type != 2)
		return "unrecognized log type";

	mad_dump_cc_congestionlog(buf, sizeof buf, data, sizeof data);

	printf("# CongestionLog: %s\n%s", portid2str(dest), buf);

	if (type == 1) {
		mad_dump_cc_congestionlogswitch(buf, sizeof buf, data, sizeof data);
		printf("%s\n", buf);
		for (i = 0; i < 15; i++) {
			/* output only if entry not 0 */
			if (memcmp(data + 40 + i * 12, emptybuf, 12)) {
				mad_dump_cc_congestionlogentryswitch(buf, sizeof buf,
								     data + 40 + i * 12,
								     12);
				printf("%s\n", buf);
			}
		}
	}
	else {
		/* XXX: Q3/2010 errata lists first entry offset at 80, but we assume
		 * will be updated to 96 once CurrentTimeStamp field is word aligned.
		 * In addition, assume max 13 log events instead of 16.  Due to 
		 * errata changes increasing size of CA log event, 16 log events is
		 * no longer possible to fit in max MAD size.
		 */
		mad_dump_cc_congestionlogca(buf, sizeof buf, data, sizeof data);
		printf("%s\n", buf);
		for (i = 0; i < 13; i++) {
			/* output only if entry not 0 */
			if (memcmp(data + 12 + i * 16, emptybuf, 16)) {
				mad_dump_cc_congestionlogentryca(buf, sizeof buf,
								 data + 12 + i * 16,
								 16);
				printf("%s\n", buf);
			}
		}
	}

	return NULL;
}

static char *switch_congestion_setting(ib_portid_t * dest, char **argv, int argc)
{
	char buf[2048];
	char data[IB_CC_DATA_SZ] = { 0 };
	
	if (!cc_query_status_via(data, dest, IB_CC_ATTR_SWITCH_CONGESTION_SETTING,
				 0, 0, NULL, srcport, cckey))
		return "switch congestion setting query failed";

	mad_dump_cc_switchcongestionsetting(buf, sizeof buf, data, sizeof data);

	printf("# SwitchCongestionSetting: %s\n%s", portid2str(dest), buf);
	return NULL;
}

static char *switch_port_congestion_setting(ib_portid_t * dest, char **argv, int argc)
{
	char buf[2048];
	char data[IB_CC_DATA_SZ] = { 0 };
	int type, numports, maxblocks, i, j;
	int portnum = 0;
	int outputcount = 0;

	if (argc > 0)
		portnum = strtol(argv[0], 0, 0);

	/* Figure out number of ports first */
	if (!smp_query_via(data, dest, IB_ATTR_NODE_INFO, 0, 0, srcport))
		return "node info query failed";

	mad_decode_field((uint8_t *)data, IB_NODE_TYPE_F, &type);
	mad_decode_field((uint8_t *)data, IB_NODE_NPORTS_F, &numports);

	if (type != IB_NODE_SWITCH)
		return "destination not a switch";

	printf("# SwitchPortCongestionSetting: %s\n", portid2str(dest));

	if (portnum) {
		if (portnum > numports)
			return "invalid port number specified";

		memset(data, '\0', sizeof data);
		if (!cc_query_status_via(data, dest, IB_CC_ATTR_SWITCH_PORT_CONGESTION_SETTING,
					 (portnum - 1) / 32, 0, NULL, srcport, cckey))
			return "switch congestion setting query failed";

		mad_dump_cc_switchportcongestionsettingelement(buf, sizeof buf,
							       data + (((portnum % 32) - 1) * 4),
							       4);
		printf("%s", buf);
		return NULL;
	}

	/* else get all port info */

	maxblocks = ((numports - 1) / 32) + 1;

	for (i = 0; i < maxblocks; i++) {
		memset(data, '\0', sizeof data);
		if (!cc_query_status_via(data, dest, IB_CC_ATTR_SWITCH_PORT_CONGESTION_SETTING,
					 i, 0, NULL, srcport, cckey))
			return "switch congestion setting query failed";

		for (j = 0; j < 32 && outputcount < numports; j++) {
			printf("Port:............................%u\n", i * 32 + j + 1);
			mad_dump_cc_switchportcongestionsettingelement(buf, sizeof buf,
								       data + j * 4,
								       4);
			printf("%s\n", buf);
			outputcount++;
		}
	}

	return NULL;
}

static char *ca_congestion_setting(ib_portid_t * dest, char **argv, int argc)
{
	char buf[2048];
	char data[IB_CC_DATA_SZ] = { 0 };
	int i;
	
	if (!cc_query_status_via(data, dest, IB_CC_ATTR_CA_CONGESTION_SETTING,
				 0, 0, NULL, srcport, cckey))
		return "ca congestion setting query failed";

	mad_dump_cc_cacongestionsetting(buf, sizeof buf, data, sizeof data);

	printf("# CACongestionSetting: %s\n%s\n", portid2str(dest), buf);

	for (i = 0; i < 16; i++) {
		printf("SL:..............................%u\n", i);
		mad_dump_cc_cacongestionentry(buf, sizeof buf,
					      data + 4 + i * 8,
					      8);
		printf("%s\n", buf);
	}
	return NULL;
}

static char *congestion_control_table(ib_portid_t * dest, char **argv, int argc)
{
	char buf[2048];
	char data[IB_CC_DATA_SZ] = { 0 };
	int limit, outputcount = 0;
	int i, j;
	
	if (!cc_query_status_via(data, dest, IB_CC_ATTR_CONGESTION_CONTROL_TABLE,
				 0, 0, NULL, srcport, cckey))
		return "congestion control table query failed";

	mad_decode_field((uint8_t *)data, IB_CC_CONGESTION_CONTROL_TABLE_CCTI_LIMIT_F, &limit);

	mad_dump_cc_congestioncontroltable(buf, sizeof buf, data, sizeof data);

	printf("# CongestionControlTable: %s\n%s\n", portid2str(dest), buf);

	if (!limit)
		return NULL;

	for (i = 0; i < (limit/64) + 1; i++) {

		/* first query done */
		if (i)
			if (!cc_query_status_via(data, dest, IB_CC_ATTR_CONGESTION_CONTROL_TABLE,
					  i, 0, NULL, srcport, cckey))
				return "congestion control table query failed";

		for (j = 0; j < 64 && outputcount <= limit; j++) {
			printf("Entry:...........................%u\n", i*64 + j);
			mad_dump_cc_congestioncontroltableentry(buf, sizeof buf,
								data + 4 + j * 2,
								sizeof data - 4 - j * 2);
			printf("%s\n", buf);
			outputcount++;
		}
	}
	return NULL;
}

static char *timestamp_dump(ib_portid_t * dest, char **argv, int argc)
{
	char buf[2048];
	char data[IB_CC_DATA_SZ] = { 0 };

	if (!cc_query_status_via(data, dest, IB_CC_ATTR_TIMESTAMP,
				 0, 0, NULL, srcport, cckey))
		return "timestamp query failed";

	mad_dump_cc_timestamp(buf, sizeof buf, data, sizeof data);

	printf("# Timestamp: %s\n%s", portid2str(dest), buf);
	return NULL;
}

static op_fn_t *match_op(char *name)
{
	const match_rec_t *r;
	unsigned len = strlen(name);
	for (r = match_tbl; r->name; r++)
		if (!strncasecmp(r->name, name, len) ||
		    (r->alias && !strncasecmp(r->alias, name, len)))
			return r->fn;
	return NULL;
}

static int process_opt(void *context, int ch, char *optarg)
{
	switch (ch) {
	case 'c':
		cckey = (uint64_t) strtoull(optarg, 0, 0);
		break;
	default:
		return -1;
	}
	return 0;
}

int main(int argc, char **argv)
{
	char usage_args[1024];
	int mgmt_classes[4] =
	    { IB_SMI_CLASS, IB_SMI_DIRECT_CLASS, IB_SA_CLASS, IB_CC_CLASS };
	ib_portid_t portid = { 0 };
	char *err;
	op_fn_t *fn;
	const match_rec_t *r;
	int n;

	const struct ibdiag_opt opts[] = {
		{"cckey", 'c', 1, "<key>", "CC key"},
		{0}
	};
	const char *usage_examples[] = {
		"CongestionInfo 3\t\t\t# Congestion Info by lid",
		"SwitchPortCongestionSetting 3\t# Query all Switch Port Congestion Settings",
		"SwitchPortCongestionSetting 3 1\t# Query Switch Port Congestion Setting for port 1",
		NULL
	};

	n = sprintf(usage_args, "[-c key] <op> <lid|guid>\n"
		    "\nSupported ops (and aliases, case insensitive):\n");
	for (r = match_tbl; r->name; r++) {
		n += snprintf(usage_args + n, sizeof(usage_args) - n,
			      "  %s (%s) <lid|guid>%s\n", r->name,
			      r->alias ? r->alias : "",
			      r->opt_portnum ? " [<portnum>]" : "");
		if (n >= sizeof(usage_args))
			exit(-1);
	}

	ibdiag_process_opts(argc, argv, NULL, "D", opts, process_opt,
			    usage_args, usage_examples);

	argc -= optind;
	argv += optind;

	if (argc < 2)
		ibdiag_show_usage();

	if (!(fn = match_op(argv[0])))
		IBERROR("operation '%s' not supported", argv[0]);

	srcport = mad_rpc_open_port(ibd_ca, ibd_ca_port, mgmt_classes, 4);
	if (!srcport)
		IBERROR("Failed to open '%s' port '%d'", ibd_ca, ibd_ca_port);

	if (ib_resolve_portid_str_via(&portid, argv[1], ibd_dest_type,
				      ibd_sm_id, srcport) < 0)
		IBERROR("can't resolve destination %s", argv[1]);
	if ((err = fn(&portid, argv + 2, argc - 2)))
		IBERROR("operation %s: %s", argv[0], err);

	mad_rpc_close_port(srcport);
	exit(0);
}
